File: backend/app/routers/analysis.py
"""
Analysis API Router
Updated: 2025-12-05 - Využívá depot_id, route_category pro párování
"""
from typing import Optional
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, and_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.models import (
    Proof, Carrier, PriceConfig, FixRate, KmRate, DepoRate, LinehaulRate
)

router = APIRouter()


# =============================================================================
# HELPER FUNKCE PRO PÁROVÁNÍ
# =============================================================================

def find_fix_rate_for_route(
    fix_rates: list,
    route_type: str,
    route_category: Optional[str] = None,
    depot_code: Optional[str] = None
) -> Optional[float]:
    """
    Najde FIX sazbu pro daný typ trasy.
    Priorita:
    1. route_category + depot match
    2. route_type exact match
    3. route_type partial match
    """
    # Priorita 1: Podle route_category a depot
    if route_category:
        for rate in fix_rates:
            if rate.route_category == route_category:
                if depot_code and rate.depot and rate.depot.code == depot_code:
                    return float(rate.rate)
                elif not depot_code:
                    return float(rate.rate)
    
    # Priorita 2: Přesná shoda route_type
    for rate in fix_rates:
        if rate.route_type == route_type:
            return float(rate.rate)
    
    # Priorita 3: Částečná shoda
    route_upper = route_type.upper()
    for rate in fix_rates:
        rt = (rate.route_type or '').upper()
        if route_upper in rt or rt in route_upper:
            return float(rate.rate)
    
    return None


def detect_route_category_from_type(route_type: str) -> Optional[str]:
    """Detekuje kategorii z názvu route_type"""
    if not route_type:
        return None
    
    rt_upper = route_type.upper()
    
    # DIRECT ze skladu = Praha, STČ
    if 'PRAHA' in rt_upper or 'STČ' in rt_upper or 'STREDNI' in rt_upper:
        return 'DIRECT_SKLAD'
    
    # DIRECT z depa = ostatní regiony
    if any(depot in rt_upper for depot in ['VRATIMOV', 'BYDZOV', 'BYDŽOV', 'BRNO', 'BUDEJOVIC', 'RAKOVNIK']):
        return 'DIRECT_DEPO'
    
    return None


def detect_depot_code_from_type(route_type: str) -> Optional[str]:
    """Detekuje kód depa z názvu route_type"""
    if not route_type:
        return None
    
    rt_upper = route_type.upper()
    
    if 'VRATIMOV' in rt_upper:
        return 'VRATIMOV'
    elif 'BYDZOV' in rt_upper or 'BYDŽOV' in rt_upper:
        return 'NOVY_BYDZOV'
    elif 'BRNO' in rt_upper:
        return 'BRNO'
    elif 'BUDEJOVIC' in rt_upper or 'BUDĚJOVIC' in rt_upper:
        return 'CESKE_BUDEJOVICE'
    elif 'RAKOVNIK' in rt_upper or 'RAKOVNÍK' in rt_upper:
        return 'RAKOVNIK'
    
    return None


# =============================================================================
# HLAVNÍ ENDPOINT
# =============================================================================

@router.get("/proof/{proof_id}")
async def analyze_proof(proof_id: int, db: AsyncSession = Depends(get_db)):
    """
    Analyzuje proof a porovnává s ceníkem.
    AKTUALIZOVÁNO: Využívá depot_id a route_category pro přesné párování.
    """
    # Načti proof se všemi vztahy
    result = await db.execute(
        select(Proof)
        .options(
            selectinload(Proof.carrier),
            selectinload(Proof.depot),
            selectinload(Proof.route_details),
            selectinload(Proof.linehaul_details),
            selectinload(Proof.depo_details),
            selectinload(Proof.invoices),
        )
        .where(Proof.id == proof_id)
    )
    proof = result.scalar_one_or_none()
    
    if not proof:
        raise HTTPException(status_code=404, detail="Proof not found")
    
    # Načti aktivní ceník s relationships
    price_config = None
    price_result = await db.execute(
        select(PriceConfig)
        .options(
            selectinload(PriceConfig.fix_rates).selectinload(FixRate.depot),
            selectinload(PriceConfig.km_rates).selectinload(KmRate.depot),
            selectinload(PriceConfig.depo_rates).selectinload(DepoRate.depot),
            selectinload(PriceConfig.linehaul_rates),
        )
        .where(
            and_(
                PriceConfig.carrier_id == proof.carrier_id,
                PriceConfig.is_active == True,
                PriceConfig.valid_from <= proof.period_date
            )
        )
        .order_by(PriceConfig.valid_from.desc())
    )
    price_config = price_result.scalars().first()
    
    # Připrav mapy sazeb
    fix_rates_map = {}
    km_rate_config = None
    depo_rates_map = {}
    linehaul_rates_map = {}
    
    if price_config:
        # FIX rates - klíč je route_type, ale ukládáme i category a depot
        for r in price_config.fix_rates:
            fix_rates_map[r.route_type] = {
                'rate': float(r.rate),
                'route_category': r.route_category,
                'depot_code': r.depot.code if r.depot else None,
            }
        
        # KM rate
        if price_config.km_rates:
            km_rate_config = float(price_config.km_rates[0].rate)
        
        # Depo rates
        for r in price_config.depo_rates:
            key = f"{r.depo_name}_{r.rate_type}"
            depo_rates_map[key] = {
                'rate': float(r.rate),
                'depot_code': r.depot.code if r.depot else None,
            }
        
        # Linehaul rates
        for r in price_config.linehaul_rates:
            key = f"{r.from_code or ''}_{r.to_code or ''}_{r.vehicle_type}"
            linehaul_rates_map[key] = float(r.rate)
    
    # === ANALÝZA ROUTE DETAILS ===
    route_details = []
    total_routes_calculated = Decimal('0')
    missing_rates = []
    
    for route in proof.route_details:
        # Detekce category a depot z route_type
        route_category = detect_route_category_from_type(route.route_type)
        depot_code = detect_depot_code_from_type(route.route_type)
        
        # Hledání sazby
        config_rate = None
        
        # 1. Přesná shoda
        if route.route_type in fix_rates_map:
            config_rate = fix_rates_map[route.route_type]['rate']
        else:
            # 2. Podle category a depot
            for rt, rate_info in fix_rates_map.items():
                if rate_info.get('route_category') == route_category:
                    if depot_code and rate_info.get('depot_code') == depot_code:
                        config_rate = rate_info['rate']
                        break
                    elif not depot_code:
                        config_rate = rate_info['rate']
                        break
        
        # 3. Fallback - partial match
        if config_rate is None:
            for rt, rate_info in fix_rates_map.items():
                if route.route_type.upper() in rt.upper() or rt.upper() in route.route_type.upper():
                    config_rate = rate_info['rate']
                    break
        
        calculated = route.count * route.rate
        total_routes_calculated += calculated
        
        status = 'ok'
        if config_rate is None:
            status = 'missing'
            missing_rates.append(f"FIX: {route.route_type}")
        elif abs(float(route.rate) - config_rate) > 1:
            status = 'warning'
        
        route_details.append({
            'routeType': route.route_type,
            'routeCategory': route_category,
            'depotCode': depot_code,
            'count': route.count,
            'proofRate': float(route.rate),
            'configRate': config_rate,
            'amount': float(route.amount),
            'calculatedAmount': float(calculated),
            'difference': float(route.amount - calculated) if route.amount else 0,
            'status': status
        })
    
    # === ANALÝZA DEPO DETAILS ===
    depo_details = []
    total_depo_calculated = Decimal('0')
    
    for depo in proof.depo_details:
        config_key = f"{depo.depo_name}_{depo.rate_type}"
        config_info = depo_rates_map.get(config_key)
        config_rate = config_info['rate'] if config_info else None
        
        calculated = (depo.days or 1) * depo.rate
        total_depo_calculated += calculated
        
        status = 'ok' if config_rate else 'missing'
        if not config_rate:
            missing_rates.append(f"DEPO: {depo.depo_name} ({depo.rate_type})")
        
        depo_details.append({
            'depoName': depo.depo_name,
            'rateType': depo.rate_type,
            'days': depo.days,
            'proofRate': float(depo.rate),
            'configRate': config_rate,
            'amount': float(depo.amount),
            'calculatedAmount': float(calculated),
            'depotCode': config_info.get('depot_code') if config_info else None,
            'status': status
        })
    
    # === ANALÝZA LINEHAUL DETAILS ===
    linehaul_details = []
    total_linehaul_calculated = Decimal('0')
    
    for lh in proof.linehaul_details:
        key = f"{lh.from_code or ''}_{lh.to_code or ''}_{lh.vehicle_type or ''}"
        config_rate = linehaul_rates_map.get(key)
        
        calculated = lh.total or (lh.rate * (lh.days or 1) * (lh.per_day or 1))
        total_linehaul_calculated += calculated
        
        if not config_rate:
            missing_rates.append(f"LH: {lh.from_code} → {lh.to_code} ({lh.vehicle_type})")
        
        linehaul_details.append({
            'description': lh.description,
            'fromCode': lh.from_code,
            'toCode': lh.to_code,
            'vehicleType': lh.vehicle_type,
            'days': lh.days,
            'perDay': lh.per_day,
            'rate': float(lh.rate),
            'total': float(lh.total),
            'calculatedTotal': float(calculated),
            'configRate': config_rate,
            'status': 'ok' if config_rate else 'missing'
        })
    
    # === SOUHRN ===
    summary = {
        'fix': {
            'proof': float(proof.total_fix or 0),
            'calculated': float(total_routes_calculated),
            'difference': float((proof.total_fix or 0) - total_routes_calculated)
        },
        'km': {
            'proof': float(proof.total_km or 0),
            'rate': km_rate_config,
        },
        'depo': {
            'proof': float(proof.total_depo or 0),
            'calculated': float(total_depo_calculated),
            'difference': float((proof.total_depo or 0) - total_depo_calculated)
        },
        'linehaul': {
            'proof': float(proof.total_linehaul or 0),
            'calculated': float(total_linehaul_calculated),
            'difference': float((proof.total_linehaul or 0) - total_linehaul_calculated)
        },
        'grandTotal': float(proof.grand_total or 0),
    }
    
    # === INVOICE STATUS ===
    invoice_status = []
    invoiced_totals = {'fix': 0, 'km': 0, 'linehaul': 0, 'depo': 0}
    
    for inv in proof.invoices:
        for item in inv.items:
            item_type_lower = (item.item_type or '').lower()
            for cat in ['fix', 'km', 'linehaul', 'depo']:
                if cat in item_type_lower:
                    invoiced_totals[cat] += float(item.amount or 0)
                    break
    
    for category, label in [('fix', 'FIX'), ('km', 'KM'), ('linehaul', 'Linehaul'), ('depo', 'DEPO')]:
        proof_amount = float(getattr(proof, f'total_{category}', 0) or 0)
        invoiced = invoiced_totals[category]
        
        invoice_status.append({
            'category': category,
            'label': label,
            'proofAmount': proof_amount,
            'invoicedAmount': invoiced,
            'remaining': proof_amount - invoiced,
            'status': 'ok' if abs(proof_amount - invoiced) < 100 else ('missing' if invoiced == 0 else 'partial')
        })
    
    # === CHECKS ===
    checks = []
    
    # Check 1: FIX matches
    fix_diff = abs(summary['fix']['difference'])
    checks.append({
        'name': 'FIX výpočet',
        'description': 'Součet FIX položek odpovídá celkovému FIX',
        'status': 'ok' if fix_diff < 100 else 'warning',
        'expected': summary['fix']['calculated'],
        'actual': summary['fix']['proof'],
        'difference': fix_diff
    })
    
    # Check 2: DEPO matches
    depo_diff = abs(summary['depo']['difference'])
    checks.append({
        'name': 'DEPO výpočet',
        'description': 'Součet DEPO položek odpovídá celkovému DEPO',
        'status': 'ok' if depo_diff < 100 else 'warning',
        'expected': summary['depo']['calculated'],
        'actual': summary['depo']['proof'],
        'difference': depo_diff
    })
    
    # Check 3: Has price config
    checks.append({
        'name': 'Ceník',
        'description': 'Existuje aktivní ceník pro toto období',
        'status': 'ok' if price_config else 'warning',
        'message': 'Ceník nalezen' if price_config else 'Chybí aktivní ceník'
    })
    
    # Check 4: Missing rates
    checks.append({
        'name': 'Chybějící sazby',
        'description': 'Všechny položky mají odpovídající sazbu v ceníku',
        'status': 'ok' if not missing_rates else 'warning',
        'message': 'Kompletní' if not missing_rates else f"Chybí: {len(missing_rates)} sazeb",
        'missingRates': missing_rates[:10]  # Max 10
    })
    
    # Check 5: All invoiced
    missing_invoices = [s['label'] for s in invoice_status if s['status'] == 'missing' and s['proofAmount'] > 0]
    checks.append({
        'name': 'Fakturace',
        'description': 'Všechny typy jsou vyfakturovány',
        'status': 'ok' if not missing_invoices else 'warning',
        'message': 'Kompletní' if not missing_invoices else f"Chybí: {', '.join(missing_invoices)}"
    })
    
    # Overall status
    has_error = any(c['status'] == 'error' for c in checks)
    has_warning = any(c['status'] == 'warning' for c in checks)
    overall_status = 'error' if has_error else 'warning' if has_warning else 'ok'
    
    return {
        'proofId': proof.id,
        'carrierId': proof.carrier_id,
        'carrierName': proof.carrier.name if proof.carrier else None,
        'period': proof.period,
        'periodDate': proof.period_date.isoformat() if proof.period_date else None,
        'fileName': proof.file_name,
        'status': overall_status,
        'summary': summary,
        'routeDetails': route_details,
        'depoDetails': depo_details,
        'linehaulDetails': linehaul_details,
        'invoiceStatus': invoice_status,
        'checks': checks,
        'missingRates': missing_rates,
        'hasPriceConfig': price_config is not None,
    }


@router.get("/summary")
async def get_analysis_summary(
    carrier_id: Optional[int] = Query(None),
    period: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db)
):
    """Get summary of all proofs analysis"""
    query = select(Proof).options(
        selectinload(Proof.carrier),
        selectinload(Proof.invoices),
    )
    
    filters = []
    if carrier_id:
        filters.append(Proof.carrier_id == carrier_id)
    if period:
        filters.append(Proof.period == period)
    
    if filters:
        query = query.where(and_(*filters))
    
    query = query.order_by(Proof.period_date.desc())
    
    result = await db.execute(query)
    proofs = result.scalars().all()
    
    summary = []
    for proof in proofs:
        invoiced_total = sum(
            float(item.amount or 0) 
            for inv in proof.invoices 
            for item in inv.items
        )
        
        proof_total = float(proof.grand_total or 0)
        diff = proof_total - invoiced_total
        
        status = 'ok'
        if abs(diff) > 1000:
            status = 'warning' if invoiced_total > 0 else 'missing'
        
        summary.append({
            'proofId': proof.id,
            'carrierId': proof.carrier_id,
            'carrierName': proof.carrier.name if proof.carrier else None,
            'period': proof.period,
            'proofTotal': proof_total,
            'invoicedTotal': invoiced_total,
            'difference': diff,
            'status': status,
        })
    
    return summary

File: backend/src/routes/analysis.js
const express = require('express');
const { PrismaClient } = require('@prisma/client');

const router = express.Router();
const prisma = new PrismaClient();

// POST analyze proof against price config
router.post('/proof/:proofId', async (req, res) => {
  try {
    const proofId = parseInt(req.params.proofId);
    
    // Get proof with all details
    const proof = await prisma.proof.findUnique({
      where: { id: proofId },
      include: {
        carrier: true,
        routeDetails: true,
        linehaulDetails: true,
        depoDetails: true,
        invoices: { include: { items: true } }
      }
    });
    
    if (!proof) {
      return res.status(404).json({ error: 'Proof not found' });
    }
    
    // Get active price config for this carrier and period
    const priceConfig = await prisma.priceConfig.findFirst({
      where: {
        carrierId: proof.carrierId,
        isActive: true,
        validFrom: { lte: proof.periodDate },
        OR: [
          { validTo: null },
          { validTo: { gte: proof.periodDate } }
        ]
      },
      include: {
        fixRates: true,
        kmRates: true,
        depoRates: true,
        linehaulRates: true
      },
      orderBy: { validFrom: 'desc' }
    });
    
    // Perform analysis
    const analysis = analyzeProof(proof, priceConfig);
    
    // Save analysis
    const savedAnalysis = await prisma.proofAnalysis.create({
      data: {
        proofId,
        status: analysis.status,
        errorsJson: analysis.errors,
        warningsJson: analysis.warnings,
        okJson: analysis.ok,
        fixExpected: analysis.fix.expected,
        fixActual: analysis.fix.actual,
        fixDifference: analysis.fix.difference,
        unexplainedFix: analysis.fix.unexplained,
        kmExpected: analysis.km.expected,
        kmActual: analysis.km.actual,
        kmDifference: analysis.km.difference,
        linehaulExpected: analysis.linehaul.expected,
        linehaulActual: analysis.linehaul.actual,
        linehaulDiff: analysis.linehaul.difference,
        depoExpected: analysis.depo.expected,
        depoActual: analysis.depo.actual,
        depoDifference: analysis.depo.difference,
        missingRatesJson: analysis.missingRates
      }
    });
    
    res.json({
      analysis: savedAnalysis,
      details: analysis
    });
  } catch (error) {
    console.error('Error analyzing proof:', error);
    res.status(500).json({ error: 'Failed to analyze proof' });
  }
});

// GET latest analysis for proof
router.get('/proof/:proofId', async (req, res) => {
  try {
    const analysis = await prisma.proofAnalysis.findFirst({
      where: { proofId: parseInt(req.params.proofId) },
      orderBy: { createdAt: 'desc' }
    });
    
    if (!analysis) {
      return res.status(404).json({ error: 'No analysis found for this proof' });
    }
    
    res.json(analysis);
  } catch (error) {
    console.error('Error fetching analysis:', error);
    res.status(500).json({ error: 'Failed to fetch analysis' });
  }
});

// GET dashboard summary for period
router.get('/dashboard', async (req, res) => {
  try {
    const { carrierId, period } = req.query;
    
    const where = {};
    if (carrierId) where.carrierId = parseInt(carrierId);
    if (period) where.period = period;
    
    // Get proofs with analyses
    const proofs = await prisma.proof.findMany({
      where,
      include: {
        carrier: { select: { name: true } },
        invoices: true,
        analyses: {
          orderBy: { createdAt: 'desc' },
          take: 1
        }
      },
      orderBy: { periodDate: 'desc' }
    });
    
    // Calculate summary
    const summary = proofs.map(proof => {
      const invoicedTotal = proof.invoices.reduce((sum, inv) => 
        sum + parseFloat(inv.totalWithoutVat || 0), 0
      );
      const proofTotal = parseFloat(proof.grandTotal || 0);
      const latestAnalysis = proof.analyses[0];
      
      return {
        id: proof.id,
        carrier: proof.carrier.name,
        period: proof.period,
        proofTotal,
        invoicedTotal,
        invoiceCount: proof.invoices.length,
        remainingToInvoice: proofTotal - invoicedTotal,
        status: latestAnalysis?.status || 'pending',
        errors: latestAnalysis?.errorsJson?.length || 0,
        warnings: latestAnalysis?.warningsJson?.length || 0
      };
    });
    
    res.json(summary);
  } catch (error) {
    console.error('Error fetching dashboard:', error);
    res.status(500).json({ error: 'Failed to fetch dashboard' });
  }
});

// Helper function to analyze proof
function analyzeProof(proof, priceConfig) {
  const errors = [];
  const warnings = [];
  const ok = [];
  const missingRates = [];
  
  const result = {
    status: 'ok',
    errors,
    warnings,
    ok,
    missingRates,
    fix: { expected: 0, actual: 0, difference: 0, unexplained: 0 },
    km: { expected: 0, actual: 0, difference: 0 },
    linehaul: { expected: 0, actual: 0, difference: 0 },
    depo: { expected: 0, actual: 0, difference: 0 }
  };
  
  if (!priceConfig) {
    warnings.push('Žádný aktivní ceník pro toto období');
    result.status = 'warning';
    return result;
  }
  
  // Analyze FIX
  let expectedFix = 0;
  const fixRatesMap = new Map(priceConfig.fixRates.map(r => [r.routeType, parseFloat(r.rate)]));
  
  for (const route of proof.routeDetails) {
    const configRate = fixRatesMap.get(route.routeType);
    if (configRate) {
      expectedFix += route.count * configRate;
    } else {
      missingRates.push({
        type: 'fix',
        routeType: route.routeType,
        proofRate: parseFloat(route.rate)
      });
      expectedFix += parseFloat(route.amount);
    }
  }
  
  result.fix.expected = expectedFix;
  result.fix.actual = parseFloat(proof.totalFix || 0);
  result.fix.difference = result.fix.actual - result.fix.expected;
  
  if (Math.abs(result.fix.difference) > 100) {
    result.fix.unexplained = result.fix.difference;
    warnings.push(`Nevysvětlený rozdíl u FIX: ${result.fix.difference.toLocaleString('cs-CZ')} Kč`);
  } else {
    ok.push('FIX: Hodnoty sedí');
  }
  
  // Analyze KM
  const kmRate = priceConfig.kmRates[0]?.rate || 10.97;
  result.km.expected = parseFloat(proof.totalKmCount || 0) * parseFloat(kmRate);
  result.km.actual = parseFloat(proof.totalKm || 0);
  result.km.difference = result.km.actual - result.km.expected;
  
  if (Math.abs(result.km.difference) > 100) {
    errors.push(`KM: Rozdíl ${result.km.difference.toLocaleString('cs-CZ')} Kč`);
  } else {
    ok.push('KM: Hodnoty sedí');
  }
  
  // Analyze Linehaul
  result.linehaul.actual = parseFloat(proof.totalLinehaul || 0);
  // Detailed linehaul analysis would go here
  
  // Analyze Depo
  result.depo.actual = parseFloat(proof.totalDepo || 0);
  // Detailed depo analysis would go here
  
  // Check invoices
  const invoicedTypes = new Set();
  proof.invoices.forEach(inv => {
    inv.items.forEach(item => invoicedTypes.add(item.itemType));
  });
  
  const requiredTypes = ['fix', 'km', 'linehaul', 'depo'];
  requiredTypes.forEach(type => {
    if (!invoicedTypes.has(type) && parseFloat(proof[`total${type.charAt(0).toUpperCase() + type.slice(1)}`] || 0) > 0) {
      warnings.push(`Chybí faktura: ${type.toUpperCase()}`);
    }
  });
  
  // Set overall status
  if (errors.length > 0) {
    result.status = 'error';
  } else if (warnings.length > 0) {
    result.status = 'warning';
  }
  
  return result;
}

module.exports = router;
